<a href="https://colab.research.google.com/github/Penpitcha0634/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import os

In [3]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/BSC_DPDM23/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [6]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [7]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

# แปลง Data ให้อยู่ในรูปแบบ Transaction data ด้วย $groupby.apply(list)$

In [8]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [9]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [10]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list) #เอาตัวแปรมาเก็บ

เตรียม data เสร็จแล้ว

# ทำ Association Rules

In [11]:
!pip install apyori #ต้อง install แพ๊กเกตใหม่ เพราะไม่ใช่ทั่วๆไปที่คนใช้ install ครั้งเดียวพอ

In [12]:
from apyori import apriori

In [13]:
apriori(transactions,min_support = 0.065)

<generator object apriori at 0x7858c4602f10>

In [14]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [15]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [16]:
len(frequentItemsets)

4

In [17]:
frequentItemsets[0][0]

frozenset({'22423'})

In [18]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [19]:
frequentItemsets[0][0].__len__()

1

In [20]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])

In [21]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

## Quiz 22386 และ 85099B คือสินค้าอะไร

In [22]:
set(data[data['StockCode']=='22386' ]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [23]:
set(data[data['StockCode']=='85099B' ]['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

## HW 1 หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ+เดือน

สั่งวันที 3 มกราคม 2567
ส่งก่อน วันจันทร์หน้า

##เตรียมข้อมูล

In [24]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

เราต้องการแค่ เดือน/วัน/ปี ใน Column InvoiceData ไม่ได้ต้องการ เวลา

แยก เดือน/วัน/ปี ออกจาก เวลา ด้วยการทำ Prepare Data

## ทำ Data Preparation

Data Preparation คือ การแปลง(Transform)ข้อมูลให้อยู่ในรูปแบบที่เหมาะสม


In [25]:
data['InvoiceDate'][0].split()[0]

'12/01/2010'

In [26]:
date = []
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)
#set(L)

In [27]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

In [28]:
# Test
date[0].split('/')

['12', '01', '2010']

## นำ Data ที่ผ่านการแปลงแล้วไปใช้

>



In [29]:
data_test_nonclean = data

In [30]:
data_test_nonclean['date'] = date
data_test_nonclean['day'] = day
data_test_nonclean['month'] = month
data_test_nonclean['year'] = year

# data_test_nonclean >>> ข้อมูลยังไม่อยู่ในมาตรฐานเดียวกัน

In [31]:
set(data_test_nonclean['date'])

{'01/04/2011',
 '01/05/2011',
 '01/06/2011',
 '01/07/2011',
 '01/09/2011',
 '01/10/2011',
 '01/11/2011',
 '01/12/2011',
 '02/01/2011',
 '02/02/2011',
 '02/03/2011',
 '02/04/2011',
 '02/06/2011',
 '02/07/2011',
 '02/08/2011',
 '02/09/2011',
 '02/10/2011',
 '02/11/2011',
 '03/01/2011',
 '03/02/2011',
 '03/03/2011',
 '03/04/2011',
 '03/06/2011',
 '03/07/2011',
 '03/08/2011',
 '03/09/2011',
 '03/10/2011',
 '03/11/2011',
 '04/01/2011',
 '04/03/2011',
 '04/04/2011',
 '04/05/2011',
 '04/06/2011',
 '04/07/2011',
 '04/08/2011',
 '04/10/2011',
 '04/11/2011',
 '04/12/2011',
 '05/01/2011',
 '05/03/2011',
 '05/04/2011',
 '05/05/2011',
 '05/06/2011',
 '05/08/2011',
 '05/09/2011',
 '05/10/2011',
 '05/11/2011',
 '05/12/2011',
 '06/01/2011',
 '06/02/2011',
 '06/03/2011',
 '06/05/2011',
 '06/06/2011',
 '06/07/2011',
 '06/08/2011',
 '06/09/2011',
 '06/10/2011',
 '06/12/2011',
 '07/01/2011',
 '07/03/2011',
 '07/04/2011',
 '07/05/2011',
 '07/06/2011',
 '07/07/2011',
 '07/08/2011',
 '07/10/2011',
 '07/11/20

In [32]:
set(data_test_nonclean['day'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31'}

### แปลงวัน ให้เป็นมาตรฐานเดียวกัน

วัน เป็นมาตรฐานเดียวกันไม่ต้องแปลง

In [33]:
data_test_nonclean.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date day  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010  01   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010  01   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   

  month  year  
0    12  2010  
1    12  2010  
2    12  2010  
3    12  2010  
4    12  2010

### แปลงเดือน ให้เป็นมาตรฐานเดียวกัน

In [34]:
set(data_test_nonclean['month'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '1',
 '10',
 '11',
 '12',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9'}

In [35]:
data_test = data_test_nonclean.replace(to_replace=['1', '2','3','4','5','6','7','8','9','10','11','12'],value=[1,2,3,4,5,6,7,8,9,10,11,12])

In [36]:
data_test.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
date           False
day            False
month          False
year           False
dtype: bool

In [37]:
data_test_nonclean[data_test_nonclean['month']=='01']

InvoiceNo StockCode                Description  Quantity  \
42481    539993     22386    JUMBO BAG PINK POLKADOT        10   
42482    539993     21499         BLUE POLKADOT WRAP        25   
42483    539993     21498        RED RETROSPOT WRAP         25   
42484    539993     22379   RECYCLING BAG RETROSPOT          5   
42485    539993     20718  RED RETROSPOT SHOPPER BAG        10   
...         ...       ...                        ...       ...   
55385    540992     21912   VINTAGE SNAKES & LADDERS         2   
55386    540992     51008   AFGHAN SLIPPER SOCK PAIR        10   
55387    540992     21508    VINTAGE KID DOLLY CARD         12   
55388    540992     22716         CARD CIRCUS PARADE        12   
55389    540992     22983        CARD BILLBOARD FONT        12   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481  01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484  01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485  01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                 ...        ...         ...             ...         ...   
55385  01/12/2011 16:57       3.75     17364.0  United Kingdom  01/12/2011   
55386  01/12/2011 16:57       3.45     17364.0  United Kingdom  01/12/2011   
55387  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55388  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55389  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   

      day month  year  
42481  04    01  2011  
42482  04    01  2011  
42483  04    01  2011  
42484  04    01  2011  
42485  04    01  2011  
...    ..   ...   ...  
55385  12    01  2011  
55386  12    01  2011  
55387  12    01  2011  
55388  12    01  2011  
55389  12    01  2011  

[12909 rows x 12 columns]

In [38]:
data_test[data_test['month']==1]

InvoiceNo StockCode                      Description  Quantity  \
55390    540993     22423         REGENCY CAKESTAND 3 TIER         2   
55391    540993     22776      SWEETHEART CAKESTAND 3 TIER         1   
55392    540993     21471     STRAWBERRY RAFFIA FOOD COVER         6   
55393    540993     47580            TEA TIME DES TEA COSY         6   
55394    540993     21212  PACK OF 72 RETROSPOT CAKE CASES        24   
...         ...       ...                              ...       ...   
77623    542745     22138    BAKING SET 9 PIECE RETROSPOT         13   
77624    542746     21034     REX CASH+CARRY JUMBO SHOPPER         2   
77625   C542748     22969     HOMEMADE JAM SCENTED CANDLES       -12   
77626   C542753     22649     STRAWBERRY FAIRY CAKE TEAPOT        -1   
77627    542759     22649     STRAWBERRY FAIRY CAKE TEAPOT         1   

           InvoiceDate  UnitPrice  CustomerID         Country       date day  \
55390   1/13/2011 9:23      12.75     13284.0  United Kingdom  1/13/2011  13   
55391   1/13/2011 9:23       9.95     13284.0  United Kingdom  1/13/2011  13   
55392   1/13/2011 9:23       3.75     13284.0  United Kingdom  1/13/2011  13   
55393   1/13/2011 9:23       2.55     13284.0  United Kingdom  1/13/2011  13   
55394   1/13/2011 9:23       0.55     13284.0  United Kingdom  1/13/2011  13   
...                ...        ...         ...             ...        ...  ..   
77623  1/31/2011 16:47       4.95     15881.0  United Kingdom  1/31/2011  31   
77624  1/31/2011 16:48       0.95     15881.0  United Kingdom  1/31/2011  31   
77625  1/31/2011 16:53       1.45     15881.0  United Kingdom  1/31/2011  31   
77626  1/31/2011 16:57       4.95     15881.0  United Kingdom  1/31/2011  31   
77627  1/31/2011 16:59       4.95     15881.0  United Kingdom  1/31/2011  31   

      month  year  
55390     1  2011  
55391     1  2011  
55392     1  2011  
55393     1  2011  
55394     1  2011  
...     ...   ...  
77623     1  2011  
77624     1  2011  
77625     1  2011  
77626     1  2011  
77627     1  2011  

[22238 rows x 12 columns]

###แยกประเทศออกตาม ทวีป

In [39]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

#### ทวีปยุโรป

In [40]:
EU_country = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

#### ทวีปเอเชีย

In [41]:
Asia_country = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

## Code หาชื่อสินค้า จาก Column "Description"

In [42]:
def finditem(x,y) :
  L=set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return L

## 1. หา Frequent Item Set และ Association Rules ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)

### จัดกลุ่มตาม Transaction (Invoice_No)

In [43]:
data_CM_and_NY = data_test[data_test['month']==12]

In [44]:
data_CM_and_NY

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
541904  09    12  2011  
541905  09    12  2011  
541906  09    12  2011  
541907  09    12  2011  
541908  09    12  2011  

[68006 rows x 12 columns]

In [45]:
transactionsCM_and_NY = data_CM_and_NY.groupby('InvoiceNo')['StockCode'].apply(list)

In [46]:
frequentItemsets = list(apriori(transactionsCM_and_NY,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'22111'}), support=0.07203947368421053, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22111'}), confidence=0.07203947368421053, lift=1.0)]),
 RelationRecord(items=frozenset({'22112'}), support=0.07072368421052631, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22112'}), confidence=0.07072368421052631, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),
 RelationRecord(items=frozenset({'22910'}), support=0.07171052631578947, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

In [47]:
len(frequentItemsets)

6

In [48]:
min_sup = 0.035

frequentItemsets = list(apriori(transactionsCM_and_NY,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])
yes
RelationRecord(items=frozenset({'22835', '22112'}), support=0.036513157894736845, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22112', '22835'}), confidence=0.036513157894736845, lift=1.0), OrderedStatistic(items_base=frozenset({'22112'}), items_add=frozenset({'22835'}), confidence=0.5162790697674419, lift=9.748374981944245), OrderedStatistic(items_base=frozenset({'22835'}), items_add=frozenset({'22112'}), confidence=0.68944099378882, lift=9.748

In [49]:
set(data[(data['StockCode']=='22086') | (data['StockCode']=='22910') | (data['StockCode']=='22835') | (data['StockCode']=='22112') | (data['StockCode']=='22727') | (data['StockCode']=='22726')]['Description'])

{'ALARM CLOCK BAKELIKE GREEN',
 'ALARM CLOCK BAKELIKE RED ',
 'CHOCOLATE HOT WATER BOTTLE',
 'HOT WATER BOTTLE I AM SO POORLY',
 "PAPER CHAIN KIT 50'S CHRISTMAS ",
 'PAPER CHAIN KIT VINTAGE CHRISTMAS',
 nan}

มี 3 itemset ที่ผ่าน 0.035

RelationRecord(items=frozenset({'22086', '22910'}),
support=0.047697368421052634,
<br>
<br>
**คนที่ซื้อ 22086(ALARM CLOCK BAKELIKE GREEN) กับ  22910(ALARM CLOCK BAKELIKE RED) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 4.77% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22086(ALARM CLOCK BAKELIKE GREEN) แล้วค่อยซื้อ 22910(ALARM CLOCK BAKELIKE RED) มีค่า confidence เท่ากับ 51.6% มีค่า lift เท่ากับ 7.19 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331),
<br>
<br>
**คนที่ซื้อ 22910(ALARM CLOCK BAKELIKE RED) แล้วค่อยซื้อ 22086(ALARM CLOCK BAKELIKE GREEN) มีค่า confidence ตั้งเท่ากับ 66.51% มีค่า lift เท่ากับ 7.19 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])

เสร็จข้อ 1

## 2. หา Frequent Item Set และ Association Rules ของประเทศในทวีป Europe


### จัดกลุ่มประเทศใส่ในตัวแปรทวีป Europe

In [74]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

In [75]:
Eu = ['Austria','Belgium','Channel Islands','Cyprus','Czech Republic','Denmark','EIRE','European Community',
      'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
      'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

### จัดกลุ่มตาม Transaction

In [76]:
Eu_data = data_test[data_test['Country']==""]
for i in range(len(Eu)):
  x = data_test[data_test['Country']==Eu[i]]
  Eu_data=pd.concat([x,Eu_data])

In [77]:
Eu_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
536964  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536965  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536966  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536967  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536968  12/08/2011 10:26      40.00     12358.0         Austria  12/08/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
536964  08    12  2011  
536965  08    12  2011  
536966  08    12  2011  
536967  08    12  2011  
536968  08    12  2011  

[538358 rows x 12 columns]

In [78]:
transactions_Eu = Eu_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [79]:
frequentItemsets = list(apriori(transactions_Eu,min_support = 0.085))
frequentItemsets

[RelationRecord(items=frozenset({'85123A'}), support=0.08713225292448797, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08713225292448797, lift=1.0)])]

In [80]:
len(frequentItemsets)

1

In [81]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions_Eu,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03233453810578679, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03233453810578679, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03011931133652015, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03011931133652015, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7402101241642789, lift=17.20537191045263), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7000903342366757, lift=17

In [82]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22697') | (data['StockCode']=='22699')]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

มี 3 itemset ที่ผ่าน 0.03

RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03233453810578679,
<br>
<br>
**คนที่ซื้อ 85099B(JUMBO BAG PINK POLKADOT) กับ 22386(GREEN REGENCY TEACUP AND SAUCER) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 3.23% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03233453810578679, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22386(GREEN REGENCY TEACUP AND SAUCER) แล้วค่อยซื้อ 85099B(JUMBO BAG PINK POLKADOT) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 67.75% มีค่า lift เท่ากับ 8.20 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402),
<br>
<br>
**คนที่ซื้อ 85099B(JUMBO BAG PINK POLKADOT) แล้วค่อยซื้อ 22386(GREEN REGENCY TEACUP AND SAUCER)  จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 39.15% มีค่า lift เท่ากับ 8.20 แสดงว่า มีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes

## 3.หา Frequent Item Set และ Association Rules ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)

## 4.หา Frequent Item Set และ Association Rules ของช่วงเวลา Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)

In [64]:
#เช็คประเทศในทวีปยุโรป
Eu = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

In [65]:
#ดึงประเทศในทวีปยุโรป มาจาก data_test
#data_test คือ ตัวที่เคลียร์ข้อมูลเรียบร้อย
Europe_CN = data_test[data_test['Country']==""]
for i in range(len(Eu)):
  x = data_test[data_test['Country']==Eu[i]]
  Europe_CN=pd.concat([x,Europe_CN])

In [66]:
Europe_CN #เรียกดูตาราง

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
536964  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536965  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536966  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536967  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536968  12/08/2011 10:26      40.00     12358.0         Austria  12/08/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
536964  08    12  2011  
536965  08    12  2011  
536966  08    12  2011  
536967  08    12  2011  
536968  08    12  2011  

[538358 rows x 12 columns]

In [67]:
#สร้างข้อมูลใหม่ โดยเรียกดูเดือนที่ 12 เพราะเป็นในช่วงเวลาที่ต้องการ
Europe_CN12 = Europe_CN[Europe_CN['month']==12]

In [68]:
len(set(Europe_CN['Country'])) #ทดไว้ว่าจะดูประเทศของทวีปยุโรป

24

In [69]:
set(Europe_CN['Country']) #ทดไว้ว่าจะดูประเทศของทวีปยุโรป

{'Austria',
 'Belgium',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Iceland',
 'Italy',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'Spain',
 'Sweden',
 'Switzerland',
 'United Kingdom'}

In [70]:
Europe_CN12 #เรียกดูตาราง

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
536964  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536965  12/08/2011 10:26       5.95     12358.0         Austria  12/08/2011   
536966  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536967  12/08/2011 10:26       0.39     12358.0         Austria  12/08/2011   
536968  12/08/2011 10:26      40.00     12358.0         Austria  12/08/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
536964  08    12  2011  
536965  08    12  2011  
536966  08    12  2011  
536967  08    12  2011  
536968  08    12  2011  

[67852 rows x 12 columns]

In [71]:
#จัดกลุ่มตาม Transaction (Invoice_No)
#แปลไทย จัดกลุ่มตามเลขที่ใบเสร็จการทำธุรกรรม เพื่อดูว่าลูกค้าซื้อสินค้าอะไรบ้าง โดยใช้ตารางEurope_CN12
transactionsEurope_CN12 = Europe_CN12.groupby('InvoiceNo')['StockCode'].apply(list)
transactionsEurope_CN12

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 3027, dtype: object

In [72]:
#apriori ติดตั้งมาเพื่องานนี้โดยเฉพาะ
#ความถี่ของการทำ item set จริงๆมันคื้อ frequentItemsetsEurope_CN12
#สนใจmin_sup ที่ 0.04 เพราะว่าถ้าเกิน0.04มันไม่มี
min_sup = 0.04
frequentEurope_CN12= list(apriori(transactionsEurope_CN12,min_support = min_sup))
for i in frequentEurope_CN12:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.04790221341261976, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.04790221341261976, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.165023343889778), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.165023343889778)])


In [73]:
finditem('22086', '22910') #เลือกเพราะว่า มี Confidence เยอะที่สุด

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

มี 1 itemset ที่ผ่าน 0.04

RelationRecord(items=frozenset({'22086', '22910'}), support=0.04790221341261976,
<br>
<br>
**คนที่ซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) กับ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 4.79% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.04790221341261976, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) แล้วค่อยซื้อ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) มีค่า confidence เท่ากับ 51.60% มีค่า lift เท่ากับ 7.17 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.165023343889778),
<br>
<br>
**คนที่ซื้อ 22910(PAPER CHAIN KIT VINTAGE CHRISTMAS) แล้วค่อยซื้อ 22086(PAPER CHAIN KIT 50'S CHRISTMAS) มีค่า confidence ตั้งเท่ากับ 66.51% มีค่า lift เท่ากับ 7.17 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.165023343889778)])

ข้อ 4 เสร็จแล้ว

## 5.หา Frequent Item Set และ Association Rules ของวันวาเลนไทน์

## 6.หา Frequent Item Set และ Association Rules โดยจัดกลุ่มตามประเทศ+เดือน

### จัดกลุ่มตามประเทศ+เดือน

In [50]:
transactions_count_month = data_test.groupby(['Country', 'month'])['StockCode'].apply(list)
transactions_count_month

Country      month
Australia    1        [22449, 22450, 22451, 84970L, 71459, 22429, 22...
             2        [23231, 23230, 23233, 22989, 22990, 22969, 229...
             3        [20665, 22232, 23052, 23053, 20749, 20750, 475...
             4                                                  [22990]
             5        [22457, 23176, 22109, 85014A, 22488, 21524, 23...
                                            ...                        
Unspecified  11       [22326, 20725, 22628, 22631, 84997D, 23254, 23...
             04       [20685, 22691, 48116, 21213, 21977, 21976, 212...
             05       [22953, 23078, 22619, 21888, 21890, 22620, 218...
             07       [21634, 21212, 22964, 23254, 23256, 84378, 209...
             09       [22904, 21329, 21992, 20772, 23196, 22543, 225...
Name: StockCode, Length: 419, dtype: object

In [61]:
min_sup = 0.27
frequentItemsets_count_month = list(apriori(transactions_count_month,min_support = min_sup))
for i in frequentItemsets_count_month:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'POST', '22326'}), support=0.29594272076372313, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '22326'}), confidence=0.29594272076372313, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8435374149659863, lift=1.497636342672662), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5254237288135593, lift=1.497636342672662)])
yes
RelationRecord(items=frozenset({'22423', 'POST'}), support=0.27923627684964203, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423', 'POST'}), confidence=0.27923627684964203, lift=1.0), OrderedStatistic(items_base=frozenset({'22423'}), items_add=frozenset({'POST'}), confidence=0.6842105263157896, lift=1.214763603925067), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22423'}), confidence=0.4957627118644068, lift=1.2147636039250

In [62]:
set(data[(data['StockCode']=='POST') | (data['StockCode']=='22326') ]['Description'])

{'POSTAGE', 'ROUND SNACK BOXES SET OF4 WOODLAND ', nan}

มี 2 itemset ที่ผ่าน 0.27

RelationRecord(items=frozenset({'POST', '22326'}), support=0.29594272076372313,
<br>
<br>
**คนที่ซื้อ POST(POSTAGE) กับ 22326(ROUND SNACK BOXES SET OF4 WOODLAND) จาก datasrt มีค่า confidance ทั้งหมดเท่ากับ 29.59% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '22326'}), confidence=0.29594272076372313, lift=1.0),
<br>
<br>
**คนที่ซื้อ 22326(ROUND SNACK BOXES SET OF4 WOODLAND)แล้วค่อยซื้อ POST(POSTAGE) จาก datasrt มีค่า confidance ทั้งหมดเท่ากับ 84.35% มีค่า lift เท่ากับ 1.498 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8435374149659863, lift=1.497636342672662),
<br>
<br>
**คนที่ซื้อ POST(POSTAGE)แล้วค่อยซื้อ 22326(ROUND SNACK BOXES SET OF4 WOODLAND)  จาก datasrt มีค่า confidance ทั้งหมดเท่ากับ 52.54% มีค่า lift เท่ากับ 1.498 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5254237288135593, lift=1.497636342672662)])

ข้อ 6 เสร็จแล้ว